core/domain/translation_domain.py

```diff
@@ -18,9 +18,262 @@
+        self.translatable_contents.update(
+            value.get_translatable_contents_collection().translatable_contents)
+
+
+class BaseTranslatableObject:
+    """Base class for all translatable objects which contain translatable
+    fields/objects. For example, a State is a translatable object in
+    Exploration, a Hint is a translatable object in State, and a hint_content
+    is a translatable field in Hint. So Exploration, State, and Hint all should
+    be the child class of BaseTranslatableObject.
+    """
+
+    def get_translatable_contents_collection(
+        self
+    ) -> TranslatableContentsCollection:
+        """Get all translatable fields in a translatable object.
+
+        Raises:
+            NotImplementedError. The derived child class must implement the
+                necessary logic to get all translatable fields in a
+                translatable object.
+        """
+        raise NotImplementedError('Must be implemented in subclasses.')
+
+    def get_all_contents_which_need_translations(
+        self,
+        entity_translation: EntityTranslation
+    ) -> List[TranslatableContent]:
+        """Returns a list of TranslatableContent instances which need new or
+        updated translations.
+
+        Args:
+            entity_translation: EntityTranslation. An object storing the
+                existing translations of an entity.
+
+        Returns:
+            list(TranslatableContent). Returns a list of TranslatableContent.
+        """
+        contents_which_need_translation = []
+        content_ids_for_translated_contents = (
+            entity_translation.translations.keys())
+        translatable_content_list = (
+            self.get_translatable_contents_collection()
+            .translatable_contents.values())
+
+        for translatable_content in translatable_content_list:
+            if translatable_content.content == '':
+                continue
+
+            if (
+                translatable_content.content_id not in
+                content_ids_for_translated_contents
+            ):
+                contents_which_need_translation.append(translatable_content)
+            elif (
+                entity_translation.translations[
+                translatable_content.content_id].needs_update
+            ):
+                contents_which_need_translation.append(translatable_content)
+
+        return contents_which_need_translation
+
+
+class EntityTranslation:
+    """The EntityTranslation represents an EntityTranslationsModel for a given
+    version of an entity in a given language.
+
+    Args:
+        entity_id: str. The id of the corresponding entity.
+        entity_type: TranslatableEntityType. The type
+            of the corresponding entity.
+        entity_version: str. The version of the corresponding entity.
+        language_code: str. The language code for the corresponding entity.
+        translations: dict(str, TranslatedContent). The translated content.
```

Comment:
This should explain what the keys and corresponding values are -- the description here is too vague.